numpy/_core/src/multiarray/unique.cpp

```diff
@@ -0,0 +1,190 @@
+    T* res = new T[hashset.size()];
+    auto it = hashset.begin();
+    size_t i = 0;
+    for (; it != hashset.end(); it++, i++) {
+        res[i] = *it;
+    }
+
+    // does this need to have the same lifetime as the array?
+    npy_intp dims[1] = {(npy_intp)hashset.size()};
+    PyArray_Descr *descr = PyArray_DESCR(self);
+    Py_INCREF(descr);
+    return PyArray_NewFromDescr(
+        &PyArray_Type,
+        descr,
+        1, // ndim
+        dims, // shape
+        NULL, // strides
+        res, // data
+        // This flag is needed to be able to call .sort on it.
+        NPY_ARRAY_WRITEABLE, // flags
+        NULL // obj
+    );
+}
+
+
+// this map contains the functions used for each item size.
+typedef std::function<PyObject *(PyArrayObject *)> function_type;
+std::unordered_map<int, function_type> unique_funcs = {
+    {NPY_BYTE, unique<npy_byte>},
+    {NPY_UBYTE, unique<npy_ubyte>},
+    {NPY_SHORT, unique<npy_short>},
+    {NPY_USHORT, unique<npy_ushort>},
+    {NPY_INT, unique<npy_int>},
+    {NPY_UINT, unique<npy_uint>},
+    {NPY_LONG, unique<npy_long>},
+    {NPY_ULONG, unique<npy_ulong>},
+    {NPY_LONGLONG, unique<npy_longlong>},
+    {NPY_ULONGLONG, unique<npy_ulonglong>},
+    {NPY_INT8, unique<npy_int8>},
+    {NPY_INT16, unique<npy_int16>},
+    {NPY_INT32, unique<npy_int32>},
+    {NPY_INT64, unique<npy_int64>},
+    {NPY_UINT8, unique<npy_uint8>},
+    {NPY_UINT16, unique<npy_uint16>},
+    {NPY_UINT32, unique<npy_uint32>},
+    {NPY_UINT64, unique<npy_uint64>},
+};
+
+
+extern "C" NPY_NO_EXPORT PyObject *
+array_unique(PyObject *NPY_UNUSED(dummy), PyObject *args)
+{
+    /* This is called from Python space, and expects a single numpy array as input.
+
+    It then returns a numpy array containing the unique values of the input array.
+
+    If the input array is not supported, it returns None.
+    */
+    // this is to allow grabbing the GIL before raising a python exception.
+    NPY_ALLOW_C_API_DEF;
+    PyArrayObject *self = NULL;
+    PyObject *res = NULL;
+    if (!PyArg_ParseTuple(args, "O&", PyArray_Converter, &self))
+        return NULL;
+    
+    // Making sure the DECREF is called when the function returns, with
+    // or w/o an exception
+    auto self_decref = finally([&]() { Py_XDECREF(self); });
+
+
+    try {
+        /* Handle zero-sized arrays specially */
+        if (PyArray_SIZE(self) == 0) {
+            return PyArray_NewLikeArray(
+                self,
+                NPY_ANYORDER,
+                NULL, // descr (use prototype's descr)
+                1 // subok
+            );
+        }
+
+        auto type = PyArray_TYPE(self);
+        // we only support data types present in our unique_funcs map
+        if (unique_funcs.find(type) == unique_funcs.end()) {
+            // Grab GIL before raising an exception
+            NPY_ALLOW_C_API;
+            PyErr_SetString(PyExc_NotImplementedError, "Unsupported dtype");
+            NPY_DISABLE_C_API;
```

Comment:
Here, we always have the GIL.  You'll need to add a `finally` function/destructor to the caller where you release the GIL (similar to the NpyIter cleanup, or including in that).

I realize that this comes from the ufunc code which is in a different situation here.  I am not sure that you are already releasing the GIL, but if you are not, you could do so before the `do {} while ()` loop and grab it again after).

Just a brief comment because I happened to look at it.  Will look at it more later (ping me in a while if I forgot).

Reply:
Now I explicitly release the GIL and grab it back in the compute heavy part.